nixos/modules/services/games/xonotic.nix

```diff
@@ -0,0 +1,130 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) mkEnableOption mkIf mkOption types;
+
+  cfg = config.services.xonotic;
+
+  toCfg = lib.generators.toKeyValue {
+    mkKeyValue = lib.generators.mkKeyValueDefault {
+      mkValueString = v:
+        if v == true then
+          "1"
+        else if v == false then
+          "0"
+        else if builtins.isString v then
+          ''"${v}"''
+        else
+          lib.generators.mkValueStringDefault { } v;
+    } " ";
+  };
+
+  toConfig = config: toCfg config;
+
+  writeConfig = { config, extraConfig }:
+    pkgs.writeText "server.cfg" ''
+      ${toConfig config}
+      ${extraConfig}
+    '';
+
+  serverModule = types.submodule ({ name, ... }: {
+    options = {
+      config = mkOption {
+        type = types.attrsOf (types.oneOf [ types.str types.int types.bool ]);
+        description = ''
+          server config, see
+          <link xlink:href="https://github.com/xonotic/xonotic/wiki/basic-server-configuration">https://github.com/xonotic/xonotic/wiki/basic-server-configuration</link>
+          and
+          <link xlink:href="https://xonotic.org/tools/cacs/#0a/0/">https://xonotic.org/tools/cacs/#0a/0/</link>
+        '';
+        default = {
+          hostname = name;
+          port = 26000;
+        };
+      };
+
+      extraConfig = mkOption {
+        type = types.lines;
+        description = "extra config written to server.cfg";
+        example = ''
+          duel
+        '';
+        default = "";
+      };
+
+      preStart = mkOption {
+        type = types.lines;
+        description = "command to execute before starting xonotic";
```

Comment:
It might be beneficial to explain this will run in a shell script before the service starts for some context. Additionally, naming this option to something a non developer/sysadmin would understand would be beneficial too. As it stands I feel like you need to be a NixOS developer to understand this option.

Reply:
I'm not good with writing descriptions. Any suggestions?

Reply:
@davidak has a passion for UX, so I wonder if he could suggest a better name for this option.

Reply:
thanks for recognizing that. i think UX is the most important thing for the success of a project, but i'm not a professional in that area :D

"preStart" and the description seems descriptive. mentioning that it is a shell command would also help.

i would question why it's needed at all. for installing additional maps, packaging them and adding to some list would be more reproducible and intuitive. but when the user wants to add more content, a `dataDir` option would be better